### PR TITLE
[Core] Adding compatibilty between triangles and quadrialetarls for mortar integration

### PR DESCRIPTION
**Pull request by loumalouomega**
This just add the exact integration for triangles and quads, and the corresponding mapper. The add to the contact will be done in  a future PR

### CODE REVIEW

File: kratos/tests/mortar_mapper_python_tests/test_double_curvature_integration_triangle_quadrilateral.mdpa [hunk 1]
```diff
@@ -0,0 +1,7896 @@
...
+	3432	16.9445	53.1582	48.0993
+	3435	20.6617	54.8622	50
+	3437	12.1727	48.584	48.4667
+	3440	19.187	54.4274	50
+	3442	17.8053	53.7537	50
+	3443	16.5279	52.8984	50
+	3444	13.2095	49.7053	50
+	3445	12.1501	48.5929	50
+	3446	14.2611	50.8259	50
+	3447	15.3515	51.9091	50
```
Comment (philbucher) [inline kratos/tests/mortar_mapper_python_tests/test_double_curvature_integration_triangle_quadrilateral.mdpa:1740]: Does this test have to be this huge?

File: kratos/utilities/exact_mortar_segmentation_utility.h [hunk 2]
```diff
@@ -416,79 +426,79 @@ class KRATOS_API(KRATOS_CORE) ExactMortarIntegrationUtility
...
-        return (norm_2(PointDest.Coordinates() - PointOrig.Coordinates()) < tolerance) ? true : false;
+        return (norm_2(rPointDest.Coordinates() - rPointOrig.Coordinates()) < tolerance) ? true : false;
     }
 
     /**
      * @brief This functions calculates the determinant of a 2D triangle (using points) to check if invert the order
-     * @param PointOrig1 First point
-     * @param PointOrig2 Second point
-     * @param PointOrig3 Third point
+     * @param rPointOrig1 First point
```
Comment (philbucher) [inline kratos/utilities/exact_mortar_segmentation_utility.h:489]: you would make the life of a reviewer much easier if you would do such changes in a separate PR

**Review by philbucher — COMMENTED** [review]
I will look through this in the next days since it is rather large

One global question: You are using more and more templates, which seems to be rather inconvenient from the usage side. => different geometry requires different Function-Name
Maybe consider this in the future

File: kratos/tests/mortar_mapper_python_tests/test_double_curvature_integration_triangle_quadrilateral.mdpa (hunk 1 above)
Comment (loumalouomega) [reply kratos/tests/mortar_mapper_python_tests/test_double_curvature_integration_triangle_quadrilateral.mdpa:1740]: Is the same already existing test for quad-tri, but combined. And for the tri-quad I just inverted the model part order

File: kratos/utilities/exact_mortar_segmentation_utility.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h:489]: Sorry, you are right

**Comment by loumalouomega** [comment]
Thanks @philbucher, yes I know that I am using many templates. This allow me to use bounded matrices and arrays. Maybe I will replace them in the future, but for this PR I was aiming to keep retro-compatibility

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h [hunk 3]
```diff
@@ -368,15 +377,83 @@ class DerivativesUtilities
         }
     }
 
+    /**
+     * @brief Calculates the increment of the normal and in the master condition
+     * @param rDeltaNormal The derivative of the normal
+     * @param rThisGeometry The geometry of the master side
+     * @note Hardcopied for performance
+     */
+    static inline void CalculateDeltaNormalMaster(
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:386]: maybe also someone else can have a look at this?
Seems ok but just to be sure

File: kratos/utilities/exact_mortar_segmentation_utility.h [hunk 4]
```diff
@@ -76,19 +79,26 @@ namespace Kratos {
  * @brief This utility calculates the exact integration necessary for the Mortar Conditions
  * @details The utility performs a mortar segmentation in order to obtain the exact integration of the geometry intersected
  * @author Vicente Mataix Ferrandiz
+ * @tparam TDim The dimension of work
+ * @tparam TNumNodes The number of nodes of the slave
+ * @tparam TBelong If we consider belonging of nodes or not
```
Comment (philbucher) [inline kratos/utilities/exact_mortar_segmentation_utility.h]: I know I asked this before, but I forgot what it means
maybe you can enhance the description?

File: kratos/utilities/exact_mortar_segmentation_utility.h [hunk 5]
```diff
@@ -500,87 +513,143 @@ class KRATOS_API(KRATOS_CORE) ExactMortarIntegrationUtility
...
     /**
      * @brief This function push backs the points that are inside
-     * @param PointList The intersection points
-     * @param AllInside The nodes that are already known as inside the other geometry
-     * @param ThisGeometry The geometry considered
+     * @param rPointList The intersection points
+     * @param rAllInside The nodes that are already known as inside the other geometry
+     * @param rThisGeometry The geometry considered
      */
+    template<SizeType TSizeCheck = TNumNodes>
```
Comment (philbucher) [inline kratos/utilities/exact_mortar_segmentation_utility.h:520]: you just moved this function from the cpp bcs it is templated now right?

File: kratos/utilities/exact_mortar_segmentation_utility.h [hunk 6]
```diff
@@ -500,87 +513,143 @@ class KRATOS_API(KRATOS_CORE) ExactMortarIntegrationUtility
...
-        );
+        VectorPoints& rPointList,
+        const array_1d<bool, TSizeCheck>& rAllInside,
+        GeometryPointType& rThisGeometry
+        )
+    {
+        for (IndexType i_node = 0; i_node < TSizeCheck; ++i_node) {
+            if (rAllInside[i_node]) {
+                // We check if the node already exists
+                bool add_point = true;
```
Comment (philbucher) [inline kratos/utilities/exact_mortar_segmentation_utility.h:530]: I think the following would be easier and more efficient (bcs of the break):
~~~
for (IndexType iter = 0; iter < rPointList.size(); ++iter) {
    if (CheckPoints(rThisGeometry[i_node], rPointList[iter])) {
        rPointList.push_back(rThisGeometry[i_node]);
        break;
    }
}
~~~

File: kratos/utilities/exact_mortar_segmentation_utility.h [hunk 7]
```diff
@@ -500,87 +513,143 @@ class KRATOS_API(KRATOS_CORE) ExactMortarIntegrationUtility
...
+        VectorPointsBelong& rPointList,
+        const array_1d<bool, TSizeCheck>& rAllInside,
+        GeometryPointType& rThisGeometry,
+        const PointBelongs& rThisBelongs
+        )
+    {
+        for (IndexType i_node = 0; i_node < TSizeCheck; ++i_node) {
+            if (rAllInside[i_node]) {
+                // We check if the node already exists
+                bool add_point = true;
```
Comment (philbucher) [inline kratos/utilities/exact_mortar_segmentation_utility.h:558]: similar to what I wrote above

File: kratos/utilities/exact_mortar_segmentation_utility.cpp [hunk 8]
```diff
@@ -101,7 +101,7 @@ bool ExactMortarIntegrationUtility<2, 2, false>::GetExactIntegration(
                 }
             }
         } else { // THIS IS NOT SUPPOSED TO HAPPEN
-            KRATOS_DEBUG_ERROR << "ERROR:: THIS IS NOT SUPPOSED TO HAPPEN!!!!\n" << OriginalSlaveGeometry << "\n" << OriginalMasterGeometry << std::endl;
+            KRATOS_DEBUG_ERROR << "ERROR:: THIS IS NOT SUPPOSED TO HAPPEN!!!!\n" << rOriginalSlaveGeometry << "\n" << rOriginalMasterGeometry << std::endl;
```
Comment (philbucher) [inline kratos/utilities/exact_mortar_segmentation_utility.cpp]: I think `KRATOS_ERROR` already prints sth with `Error`
just saying

**Review by philbucher — COMMENTED** [review]
puh this is really tough to review with all the cosmetic changes mixed with the real changes.
Please make two PRs next time or at least put all of them in a separate commit.
Such things will also be approved much faster

Regarding the changes themselves I have a few comments, but otherwise I approve
I am not the expert on those integration algorithms but seems ok to me

**Comment by philbucher** [comment]
Just for docu, there are quite a lot of template args now, imo this does not make it easier to be used.
Just as a hint for the future, has nothing to do with this PR

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h (hunk 3 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:386]: There is a test, the hardcopy is just to avoid the need to create a redundant template

File: kratos/utilities/exact_mortar_segmentation_utility.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h]: When you do the intersections in order to get the directional derivatives you need to know where the intersections belongs to calculate the derivatives. This says between whcih nodes the intersection belongs

File: kratos/utilities/exact_mortar_segmentation_utility.h (hunk 5 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h:520]: Yes, many errors, it was easier to move this

File: kratos/utilities/exact_mortar_segmentation_utility.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h:530]: I will check

File: kratos/utilities/exact_mortar_segmentation_utility.h (hunk 7 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h:558]: I will check

File: kratos/utilities/exact_mortar_segmentation_utility.cpp (hunk 8 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.cpp]: Ups

**Comment by loumalouomega** [comment]
I know, but it is difficult to keep the current design without adding new template arguments

File: kratos/utilities/exact_mortar_segmentation_utility.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h:530]: Doesn't work, is the opposite, we consider true by default, and we check if we don't add

File: kratos/utilities/exact_mortar_segmentation_utility.h (hunk 7 above)
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h:558]: Idem, I have not idea how to simplify this loop

**Comment by loumalouomega** [comment]
Done (what I could)

**Review by philbucher — APPROVED** [review]
ok
